File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/photoreceiver/MentraPhotoReceiverModule.kt
package com.mentra.bluetoothsdk.photoreceiver

import android.content.Context
import android.net.Uri
import android.util.Log
import com.mentra.bluetoothsdk.debug.BleTraceLogger
import com.mentra.bluetoothsdk.net.LocalIpv4
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MentraPhotoReceiverModule : Module() {
  private var photoUploadServer: LocalPhotoUploadServer? = null

  override fun definition() = ModuleDefinition {
    Name("MentraPhotoReceiver")

    Events("photoUpload", "receiverStatus")

    AsyncFunction("isSupported") {
      true
    }

    AsyncFunction("startPhotoReceiver") {
      startPhotoReceiver()
    }

    AsyncFunction("stopPhotoReceiver") {
      stopPhotoReceiverInternal()
    }

    OnDestroy {
      closePhotoReceiverInternal()
    }
  }

  @Synchronized
  private fun startPhotoReceiver(): Map<String, Any> {
    val server = photoUploadServer ?: LocalPhotoUploadServer(
      context = reactContext(),
      onLog = { message -> emitStatus(message) },
      onUpload = ::handlePhotoUpload,
    ).also {
      photoUploadServer = it
    }

    val host = LocalIpv4.bestLocalIpv4Address()
      ?: throw IllegalStateException("No Wi-Fi/LAN IPv4 address found for this phone.")

    server.activePort?.let { activePort ->
      val endpoint = receiverEndpoint(host, activePort)
      LocalPhotoReceiverRegistry.register(endpoint.uploadUrl)
      emitStatus("Photo receiver ready at ${endpoint.uploadUrl}")
      return receiverResult(endpoint)
    }

    var lastError: Throwable? = null
    for (port in PHOTO_PORTS) {
      try {
        val actualPort = server.start(port)
        val endpoint = receiverEndpoint(host, actualPort)
        LocalPhotoReceiverRegistry.register(endpoint.uploadUrl)
        emitStatus("Photo receiver ready at ${endpoint.uploadUrl}")
        return receiverResult(endpoint)
      } catch (error: Throwable) {
        lastError = error
        emitStatus("Port $port unavailable: ${error.message ?: error::class.java.simpleName}")
      }
    }

    throw IllegalStateException(
      "Could not start phone photo receiver: ${lastError?.message ?: "all ports unavailable"}",
    )
  }

  @Synchronized
  private fun stopPhotoReceiverInternal() {
    photoUploadServer?.stop()
    LocalPhotoReceiverRegistry.unregister()
    emitStatus("Photo receiver stopped")
  }

  @Synchronized
  private fun closePhotoReceiverInternal() {
    photoUploadServer?.close()
    photoUploadServer = null
    LocalPhotoReceiverRegistry.unregister()
  }

  private fun handlePhotoUpload(upload: PhotoUpload) {
    val fileUri = Uri.fromFile(upload.photoFile).toString()
    try {
      BleTraceLogger.logMap(
        "phone_to_app",
        "photo_receiver_event",
        "photo_upload",
        mapOf(
          "requestId" to upload.requestId.orEmpty(),
          "fileName" to upload.photoFile.name,
          "byteCount" to upload.byteCount,
        ),
      )
    } catch (_: Throwable) {
    }
    sendEvent(
      "photoUpload",
      mapOf(
        "requestId" to upload.requestId,
        "fileUri" to fileUri,
        "byteCount" to upload.byteCount,
      ),
    )
    emitStatus("Photo uploaded (${upload.byteCount} bytes)")
  }

  private fun emitStatus(message: String) {
    Log.d(TAG, message)
    sendEvent(
      "receiverStatus",
      mapOf("message" to message),
    )
  }

  private fun reactContext(): Context {
    return appContext.reactContext
      ?: appContext.currentActivity
      ?: throw Exceptions.ReactContextLost()
  }

  private fun receiverEndpoint(host: String, port: Int): ReceiverEndpoint {
    return ReceiverEndpoint(
      uploadUrl = "http://$host:$port/upload",
      host = host,
      port = port,
    )
  }

  private fun receiverResult(endpoint: ReceiverEndpoint): Map<String, Any> {
    return mapOf(
      "uploadUrl" to endpoint.uploadUrl,
      "host" to endpoint.host,
      "port" to endpoint.port,
    )
  }

  private companion object {
    const val TAG = "MentraPhotoReceiver"
    val PHOTO_PORTS = listOf(8787, 8788, 8789, 8790)
  }

  private data class ReceiverEndpoint(
    val uploadUrl: String,
    val host: String,
    val port: Int,
  )
}

File: mobile/modules/bluetooth-sdk/ios/MentraPhotoReceiverModule.swift
import Darwin
import ExpoModulesCore
import Foundation

public class MentraPhotoReceiverModule: Module {
  private let receiverLock = NSLock()
  private var photoUploadServer: LocalPhotoUploadServer?

  public func definition() -> ModuleDefinition {
    Name("MentraPhotoReceiver")

    Events("photoUpload", "receiverStatus")

    AsyncFunction("isSupported") {
      true
    }

    AsyncFunction("startPhotoReceiver") { () -> [String: Any] in
      try startPhotoReceiver()
    }

    AsyncFunction("stopPhotoReceiver") {
      stopPhotoReceiverInternal()
    }

    OnDestroy {
      stopPhotoReceiverInternal()
    }
  }

  private func startPhotoReceiver() throws -> [String: Any] {
    receiverLock.lock()
    defer { receiverLock.unlock() }

    guard let host = LocalIPv4.bestLocalIPv4Address() else {
      throw PhotoReceiverError("No Wi-Fi/LAN IPv4 address found for this phone.")
    }

    let server = photoUploadServer ?? LocalPhotoUploadServer(
      onLog: { [weak self] message in
        self?.emitStatus(message: message)
      },
      onUpload: { [weak self] upload in
        self?.handlePhotoUpload(upload)
      }
    )
    photoUploadServer = server

    if let activePort = server.activePort {
      let uploadUrl = "http://\(host):\(activePort)/upload"
      emitStatus(message: "Photo receiver ready at \(uploadUrl)")
      return receiverResult(uploadUrl: uploadUrl, host: host, port: activePort)
    }

    var lastError: Error?
    for port in photoPorts {
      do {
        let actualPort = try server.start(port: UInt16(port))
        let uploadUrl = "http://\(host):\(actualPort)/upload"
        emitStatus(message: "Photo receiver ready at \(uploadUrl)")
        return receiverResult(uploadUrl: uploadUrl, host: host, port: actualPort)
      } catch {
        lastError = error
        emitStatus(message: "Port \(port) unavailable: \(error.localizedDescription)")
      }
    }

    throw PhotoReceiverError(
      "Could not start phone photo receiver: \(lastError?.localizedDescription ?? "all ports unavailable")"
    )
  }

  private func stopPhotoReceiverInternal() {
    receiverLock.lock()
    defer { receiverLock.unlock() }

    photoUploadServer?.stop()
    emitStatus(message: "Photo receiver stopped")
  }

  private func handlePhotoUpload(_ upload: PhotoUpload) {
    BleTraceLogger.logMap(
      direction: "phone_to_app",
      layer: "photo_receiver_event",
      type: "photo_upload",
      payload: [
        "requestId": upload.requestId ?? "",
        "fileName": upload.photoFile.lastPathComponent,
        "byteCount": upload.byteCount,
      ]
    )
    sendEvent("photoUpload", [
      "requestId": upload.requestId as Any,
      "fileUri": upload.photoFile.absoluteString,
      "byteCount": upload.byteCount,
    ])
    emitStatus(message: "Photo uploaded (\(upload.byteCount) bytes)")
  }

  private func emitStatus(message: String) {
    NSLog("[MentraPhotoReceiver] %@", message)
    sendEvent("receiverStatus", [
      "message": message,
    ])
  }

  private func receiverResult(uploadUrl: String, host: String, port: UInt16) -> [String: Any] {
    [
      "uploadUrl": uploadUrl,
      "host": host,
      "port": port,
    ]
  }

  private let photoPorts = [8787, 8788, 8789, 8790]
}

final class PhotoReceiverError: Exception, @unchecked Sendable {
  private let message: String

  init(_ message: String) {
    self.message = message
    super.init()
  }

  override var reason: String {
    message
  }
}
